bindgen/codegen/mod.rs

```diff
@@ -2105,6 +2105,7 @@ impl CodeGenerator for CompInfo {
         let custom_derives = ctx.options().all_callbacks(|cb| {
             cb.add_derives(&crate::callbacks::DeriveInfo {
                 name: &canonical_name,
+                kind: Some(self.kind()),
```

Comment:
Continuation of https://github.com/rust-lang/rust-bindgen/pull/2328#discussion_r1043750930

So what I'd do is defining a new `enum` called `TypeKind` or `ItemKind` that can be used instead of `CompKind` inside of `DeriveInfo`:

```rust
enum WhateverKind {
    Struct,
    Enum,
    Union,
}
```

and then use the condition `is_union && struct_layout.is_rust_union()` to decide if the current type will be represented as a rust `union` or as a `struct` and use `WhateverKind::Union` or `WhateverKind::Struct` accordingly instead of `self.kind()`